### PR TITLE
fix #15973 getStacktrace works in VM

### LINKED ISSUE #15973: `getStacktrace` doesn't work in VM
`getStacktrace` doesn't work in VM, even though `doAssert false` works and shows stacktrace

### Example
```nim
proc main()=
  # doAssert false # works, shows stacktrace
  echo getStacktrace() # Error: cannot evaluate at compile time: framePtr
static: main()
```

### Current Output
```
Error: cannot evaluate at compile time: framePtr
```

### Expected Output
works

### Additional Information
* devel 1.5.1 c39fa0d49535da7f803d9972abdb15d09eed92a6
* https://github.com/nim-lang/RFCs/issues/283 would fix this

### PR DESCRIPTION
**Pull request by timotheecour**
* fix #15973 getStacktrace works in VM
* stackTraceAux now computed with iteration instead of recursion
* stackTraceAux now reuses code instead of duplicating it
* see tests/errmsgs/tproper_stacktrace4.nim

### CODE REVIEW

File: compiler/vm.nim [hunk 1]
```diff
@@ -542,6 +552,7 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
         "rc", regDescr("rc", instr.regC)]
 
     c.profiler.enter(c, tos)
+    c.tosSaved = tos
```
Comment (Araq) [inline compiler/vm.nim:555]: Yet another store in the critical path? Why?

File: compiler/vmdef.nim [hunk 2]
```diff
@@ -237,6 +237,17 @@ type
...
+  TStackFrame* = object
+    prc*: PSym                 # current prc; proc that is evaluated
+    slots*: seq[TFullReg]      # parameters passed to the proc + locals;
+                              # parameters come first
+    next*: PStackFrame         # for stacking
+    comesFrom*: int
+    safePoints*: seq[int]      # used for exception handling
+                              # XXX 'break' should perform cleanup actions
+                              # What does the C backend do for it?
+
```
Comment (Araq) [inline compiler/vmdef.nim:251]: So previously, this was an implementation detail of ``vm.nim``, now it's in ``vmdef.nim``. Seems to be a refactoring that goes in the wrong direction.

**Comment by Araq** [comment]
I seems to me that in this case a separate opcode like ``opcGetStackTrace`` would be the less intrusive change.

File: compiler/vmdef.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/vmdef.nim:251]: what do you mean? I didn't add any dependencies/imports (besides innocuous std/private/miscdollars); the only diff is adding `tosSaved*: PStackFrame` in `TCtx`
(and moving PStackFrame, TStackFrame above PCtx but that's just cosmetic and optional)

PStackFrame already was in vmdef (and used in vmprofiler.nim, vm.nim) before PR.

Or are you referring to the addition of `tosSaved` field? if so, what are concrete downsides?

Comment (Araq) [reply compiler/vmdef.nim:251]: I misread the diff, sorry.

**Comment by Araq** [comment]
I did misread the diff, but the general point remains. Shouldn't this one be an opcode?

**Pull request closed without merging**